File: web/src/app/core/api-error.ts
import { HttpErrorResponse } from '@angular/common/http';
import { FormGroup } from '@angular/forms';

/** Mirrors com.qvety.common.ApiError; error bodies are not part of the OpenAPI schema. */
export interface ApiErrorDto {
  code: string;
  message: string;
  fields?: Record<string, string>;
}

/**
 * Backend errors arrive as {code, message, fields} already localized per Accept-Language.
 * Prefer that text over a client-side key; fall back when the body is not an ApiError.
 */
export function apiError(e: unknown): ApiErrorDto | null {
  if (e instanceof HttpErrorResponse && e.error && typeof e.error === 'object' && 'code' in e.error) {
    return e.error as ApiErrorDto;
  }
  return null;
}

export function apiMessage(e: unknown, fallback: string): string {
  return apiError(e)?.message ?? fallback;
}

/** Puts per-field backend messages on the matching controls so nzErrorTip can show them. */
export function applyFieldErrors(e: unknown, form: FormGroup): boolean {
  const fields = apiError(e)?.fields;
  if (!fields) return false;
  let applied = false;
  for (const [name, message] of Object.entries(fields)) {
    const control = form.get(name);
    if (control) {
      control.setErrors({ server: message });
      control.markAsTouched();
      applied = true;
    }
  }
  return applied;
}

File: src/main/java/com/qvety/common/ApiError.java
package com.qvety.common;

import io.swagger.v3.oas.annotations.media.Schema;
import io.swagger.v3.oas.annotations.media.Schema.RequiredMode;
import java.util.Map;

/** The one error shape. code is stable and machine-readable; message is localized per Accept-Language. */
@Schema(name = "ApiError")
public record ApiError(
    @Schema(requiredMode = RequiredMode.REQUIRED) String code,
    @Schema(requiredMode = RequiredMode.REQUIRED) String message,
    Map<String, String> fields
) {}
